### ISSUE TITLE
TypeError: uint8ArrayFromString is not a function

### ISSUE THREAD

**Issue by dapplion** [libp2p/js-libp2p#967]
Lodestar broke in the docker nightly build image https://hub.docker.com/layers/chainsafe/lodestar/next/images/sha256-7b89deddd5395a3baad61aed305b5fb67e3b6ffb3fe01579b2fcb1e4d7bd821e?context=explore on start with 

```
Aug-19 10:12:21.586 []                 info: Lodestar version=v0.28.1/master/+7/eba8c7ed (git), network=pyrmont
Aug-19 10:12:21.786 [DB]               info: Connected to LevelDB database name=/data/chain-db
Aug-19 10:12:27.765 []                 info: Initializing beacon state from anchor state slot=0, epoch=0, stateRoot=0x2bb257ca66d05a047a65fe43a5f457b674de445d917cca029efb09b3ba4758c4
TypeError: uint8ArrayFromString is not a function
    at new IdentifyService (/usr/app/node_modules/libp2p/src/identify/index.js:59:66)
    at new Libp2p (/usr/app/node_modules/libp2p/src/index.js:257:30)
    at new NodejsNode (/usr/app/node_modules/@chainsafe/lodestar/src/network/nodejs/bundle.ts:93:5)
    at Object.createNodeJsLibp2p (/usr/app/node_modules/@chainsafe/lodestar/src/network/nodejs/util.ts:62:10)
    at Object.beaconHandler [as handler] (/usr/app/node_modules/@chainsafe/lodestar-cli/src/cmds/beacon/handler.ts:78:15)
```

You can recreate by just running that image with the command `beacon`

In a dockerized build this import is incorrect, adding a console log after the import it prints that
```
uint8ArrayFromString = { fromString: [Function: fromString] }
```

https://github.com/libp2p/js-libp2p/blob/9f0582f372be9c61460bad591fdb400982716be9/src/identify/index.js#L11

CC: @wemeetagain

**Comment by vasco-santos** [comment]
@dapplion can you do `npm ls uint8arrays`  on lodestar?

`libp2p` did not update ti `uint8arrays@3.0.0` yet. However, this outputs seems like `uint8arrays@3.0.0` is the used version

**Comment by mpetrunic** [comment]
> @dapplion can you do `npm ls uint8arrays` on lodestar?
> 
> `libp2p` did not update ti `uint8arrays@3.0.0` yet. However, this outputs seems like `uint8arrays@3.0.0` is the used version

@dapplion `yarn list --pattern "uint8arrays"` since we use yarn

**Comment by dapplion** [comment]
> @dapplion can you do `npm ls uint8arrays` on lodestar?
> 
> `libp2p` did not update ti `uint8arrays@3.0.0` yet. However, this outputs seems like `uint8arrays@3.0.0` is the used version

@vasco-santos 

```
$ yarn list --pattern "uint8arrays"
yarn list v1.22.5
├─ @chainsafe/lodestar@0.28.1
│  ├─ libp2p-crypto@0.19.0
│  │  └─ uint8arrays@1.1.0
│  └─ uint8arrays@2.0.5
├─ cids@1.1.5
│  ├─ multicodec@2.1.0
│  │  └─ uint8arrays@1.1.0
│  └─ uint8arrays@2.0.5
├─ libp2p-interfaces@0.8.3
│  ├─ libp2p-crypto@0.19.0
│  │  └─ uint8arrays@1.1.0
│  └─ uint8arrays@2.0.5
├─ libp2p-noise@3.1.0
│  ├─ libp2p-crypto@0.19.0
│  │  └─ uint8arrays@1.1.0
│  └─ uint8arrays@2.0.5
├─ multicodec@2.0.1
│  └─ uint8arrays@1.0.0
├─ multihashes@3.1.2
│  └─ uint8arrays@2.0.5
└─ uint8arrays@1.1.0
Done in 1.08s.
```

**Comment by dapplion** [comment]
Note that when running from source it does not break, it only breaks dockerized.

Our Dockerfile only installs
```
RUN npm install @chainsafe/lodestar-cli@$VERSION
``` 
https://github.com/ChainSafe/lodestar/blob/master/Dockerfile

**Comment by vasco-santos** [comment]
Interesting, this looks like something is not doing a good job on dependency resolution. The breaking change in `uint8arrays@3` makes the named exports as your docker output. But, `libp2p` is locked with 2 https://github.com/libp2p/js-libp2p/blob/master/package.json#L127

**Comment by vasco-santos** [comment]
I can ship https://github.com/libp2p/js-libp2p/pull/965 and it should solve the "issue". But, something is wrong somewhere else

**Comment by dapplion** [comment]
> I can ship #965 and it should solve the "issue". But, something is wrong somewhere else

@vasco-santos Adding require.resolve, the dependency is from `/root/node_modules/libp2p/node_modules/uint8arrays/cjs/src/from-string.js`, which indeed it's 3.0.0. libp2p version from root node_modules is 0.30.12

**Comment by dapplion** [comment]
I can recreate in a new container just by pulling the node image, and installing
```
npm init -y && npm i @chainsafe/lodestar-cli@next
```

**Comment by vasco-santos** [comment]
https://github.com/libp2p/js-libp2p/blob/v0.30.12/package.json#L56 shows that we do not have an `uint8array` dependency but we used it https://github.com/libp2p/js-libp2p/blob/v0.30.12/src/identify/index.js#L12

So, it is probably getting `uint8arrays` from any other dependency which already updated. Bah, I can publish a `libp2p@0.30.13` adding uint8arrays to the package.json

**Comment by dapplion** [comment]
@vasco-santos Please not sure how to unbreak out dockers otherwise

**Comment by vasco-santos** [comment]
@dapplion on it. By the way, is there anything blocking you from updating `libp2p`?

**Comment by vasco-santos** [comment]
`0.30.13` is out

**Comment by dapplion** [comment]
> @dapplion on it. By the way, is there anything blocking you from updating `libp2p`?

Yes, but I have to check-in with @wemeetagain

**Comment by wemeetagain** [comment]
> By the way, is there anything blocking you from updating libp2p?

I need to write up an issue. But there's an typescript issue between our version of `MuxedStream` and libp2p's version.

We're using a type like:
```ts
export type Libp2pStream = {
  source: AsyncIterable<Buffer>;
  sink: (source: AsyncIterable<Buffer>) => Promise<void>;
  /**
   * `libp2p-mplex`: Close for reading
   * ```ts
   * () => stream.source.end()
   * ```
   */
  close: () => void;
  /**
   * `libp2p-mplex`: Close immediately for reading and writing (remote error)
   */
  reset: () => void;
  /**
   * `libp2p-mplex`: Close for reading and writing (local error)
   */
  abort: (err: Error) => void;
};
```

Libp2p has a type that looks like:
```ts
export interface MuxedStream extends AsyncIterable<Uint8Array | BufferList> {
  source: AsyncIterable<Uint8Array | BufferList>;
  sink:  (source: Uint8Array) => Promise<void>;
  close: () => void;
  reset: () => void;
  abort: () => void;
  timeline: MuxedTimeline;
  id: string;
}
```

The real issues are `source`, `sink`. It's strange because even with the type issues, the code seems to work without runtime issues.

**Comment by dapplion** [comment]
I confirm that the new libp2p release fixes the issue in our dockerized setup. Thank you @vasco-santos :+1:

**Comment by dapplion** [comment]
@vasco-santos I'm not sure how to prevent this from happening in the future. If you install any lodestar package on its own it will break because the dependencies are de-duped.

Env with yarn.lock
```
─┬ libp2p@0.31.7
    ├─┬ multicodec@3.0.1
    │ └── uint8arrays@2.1.10
    ├─┬ multihashing-async@2.1.2
    │ └── uint8arrays@2.1.10
    └─┬ multistream-select@2.0.0
      └── uint8arrays@2.1.5 deduped
```
Installing from NPM
```
+-- libp2p@0.31.7
  | | +-- multicodec@3.2.0
  | | | `-- uint8arrays@3.0.0  deduped
  | | +-- multihashing-async@2.1.4
  | | | `-- uint8arrays@3.0.0  deduped
  | | `-- multistream-select@2.0.1
  | |   `-- uint8arrays@3.0.0  deduped
```

I would ask you to not do a breaking change on a sub-dependency while just doing a patch on the parent. But that doesn't seem a sustainable solution. How can we approach this issue so Lodestar doesn't break on this type of situations? Now I'll do a plumbing campaign to update everything.

**Comment by tuyennhv** [comment]
@vasco-santos @dapplion 
right now libp2p@0.31.7 declares `uint8arrays` as `devDependencies` so it's inconsistent for an application to decide which version of `uint8arrays` to install when it wants to use libp2p (proved by`yarn why` or `npm ls`), libp2p should declare it as `dependencies` instead https://github.com/libp2p/js-libp2p/blob/f7183e8afd49deb40d03fca708a3070c1f16610f/package.json#L168

note that we can't find "@chainsafe#lodestar-cli#@chainsafe#lodestar#libp2p#uint8arrays" in the list below using `yarn why uint8arrays`
```
=> Found "uint8arrays@3.0.0"
info Has been hoisted to "uint8arrays"
info Reasons this module exists
   - Hoisted from "peer-id#cids#uint8arrays"
   - Hoisted from "peer-id#libp2p-crypto#uint8arrays"
   - Hoisted from "peer-id#multihashes#uint8arrays"
   - Hoisted from "@chainsafe#lodestar-cli#@chainsafe#lodestar#@chainsafe#libp2p-noise#uint8arrays"
   - Hoisted from "peer-id#cids#multicodec#uint8arrays"
   - Hoisted from "@chainsafe#lodestar-cli#@chainsafe#lodestar#libp2p#multihashing-async#uint8arrays"
   - Hoisted from "@chainsafe#lodestar-cli#@chainsafe#lodestar#libp2p#multistream-select#uint8arrays"
   - Hoisted from "@chainsafe#lodestar-cli#@chainsafe#lodestar#@chainsafe#libp2p-noise#peer-id#uint8arrays"
   - Hoisted from "@chainsafe#lodestar-cli#@chainsafe#lodestar#@types#datastore-level#datastore-level#datastore-core#uint8arrays"
```

seems like latter versions of libp2p fixed this, at least master declares `uint8arrays` as `dependencies` and migrates to 3.0.0

**Comment by vasco-santos** [comment]
I added hot fix for `0.30.x` line to get this fixed there. Not for `0.31.x` as we talked about updating to all the new stuff aka `0.32.x`.

Anyway, I just added also a hot fix for in `libp2p@0.31.8` that fixes this by putting uint8arrays as a dependency, as it should be

**Comment by lidel** [comment]
@vasco-santos does this mean this issue can be closed?
